internal/loader/configuration_client_manager.go

```diff
@@ -0,0 +1,424 @@
+	manager := &ConfigurationClientManager{
+		ReplicaDiscoveryEnabled: provider.Spec.ReplicaDiscovery,
+		ValidDomain:             getValidDomain(*provider.Spec.Endpoint),
+		Endpoint:                *provider.Spec.Endpoint,
+		ClientsFallbackStatus:   make(map[string]*ConfigurationClientWrapper),
+	}
+
+	var err error
+	if provider.Spec.ConnectionStringReference != nil {
+		manager.ConnectionString, err = getConnectionStringParameter(ctx, types.NamespacedName{Namespace: provider.Namespace, Name: *provider.Spec.ConnectionStringReference})
+		if err != nil {
+			return nil, err
+		}
+		manager.secret, err = parseConnectionString(manager.ConnectionString, SecretSection)
+		if err != nil {
+			return nil, err
+		}
+		manager.id, err = parseConnectionString(manager.ConnectionString, IdSection)
+		if err != nil {
+			return nil, err
+		}
+
+		manager.SpecifiedClient, err = azappconfig.NewClientFromConnectionString(manager.ConnectionString, clientOptionWithModuleInfo)
+		if err != nil {
+			return nil, err
+		}
+
+		return manager, nil
+	}
+
+	manager.Credential, err = CreateTokenCredential(ctx, provider.Spec.Auth, provider.Namespace)
+	if err != nil {
+		return nil, err
+	}
+
+	manager.SpecifiedClient, err = azappconfig.NewClient(*provider.Spec.Endpoint, manager.Credential, clientOptionWithModuleInfo)
+	if err != nil {
+		return nil, err
+	}
+
+	return manager, nil
+}
+
+func SrvTargetHostQuery(host string) ([]string, error) {
+	results := make([]string, 0)
+
+	_, originRecords, err := net.LookupSRV(Origin, TCP, host)
+	if err != nil {
+		// If the host does not have SRV records => no replicas
+		if _, ok := err.(*net.DNSError); ok {
+			results = append(results, host)
+			return results, nil
+		}
+		return nil, err
+	}
+
+	originHost := strings.TrimSuffix(originRecords[0].Target, ".")
+	results = append(results, originHost)
+	index := 0
+	for {
+		currentAlt := Alt + strconv.Itoa(index)
+		_, altRecords, err := net.LookupSRV(currentAlt, TCP, originHost)
+		if err != nil {
+			// If the host does not have SRV records => no more replicas
+			if _, ok := err.(*net.DNSError); ok {
+				break
+			} else {
+				return nil, err
+			}
+		}
+
+		altHost := strings.TrimSuffix(altRecords[0].Target, ".")
+		if altHost != "" {
+			results = append(results, altHost)
+		}
+		index = index + 1
+	}
+
+	return results, nil
+}
+
+func (manager *ConfigurationClientManager) GetClients() []*ConfigurationClientWrapper {
+	if manager.ReplicaDiscoveryEnabled != true {
+		return []*ConfigurationClientWrapper{{
+			Endpoint:       manager.Endpoint,
+			Client:         manager.SpecifiedClient,
+			BackOffEndTime: metav1.Time{},
+			FailedAttempts: 0,
+		}}
+	}
+
+	clients := make([]*ConfigurationClientWrapper, 0)
+	currentTime := metav1.Now()
+	if currentTime.After(manager.lastFallbackClientsAttempt.Time.Add(MinimalClientRefreshInterval)) &&
+		(manager.DynamicClients == nil ||
+			currentTime.After(manager.lastFallbackRefresh.Time.Add(FallbackClientRefreshExpireInterval))) {
```

Comment:
what is `lastFallbackClientsAttempt` time and `lastFallbackRefresh` time?

Reply:
lastFallbackRefresh -> lastFallbackClientRefresh
lastFallbackClientsRefreshAttempt -> lastFallbackClientRefreshAttempt

Reply:
`lastFallbackClientsAttempt` time is the moment provider try to query srv hosts. It may fail during query process, if the query succeed, `lastFallbackRefresh` time will be updated.

Reply:
We use them to ensure two things:
1. Not perform the lookup too frequently, if we see the last attempt is within a certain time period (30s), we will not perform a new lookup
2. To ensure discovered clients could be refreshed periodically (every 1 hour), if the last success time exceeds 1 hour, we perform a new lookup